federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPMappersComparator.java

```diff
@@ -98,13 +93,21 @@ public int compare(ComponentModel o1, ComponentModel o2) {
                 if (isO2LdapAttr) {
                     return 1;
                 } else {
-                    return 0;
+                    return compareWithStableOrdering(o1, o2);
                 }
             } else if (!isO2LdapAttr) {
                 return -1;
             }
 
-            return 0;
+            return compareWithStableOrdering(o1, o2);
+        }
+
+        /**
+         * Ensure a stable ordering, so the mappers are always executed in the same order.
+         * This can avoid database deadlocks as the mappers will modify attributes always in the same order.
+         */
+        private static int compareWithStableOrdering(ComponentModel o1, ComponentModel o2) {
+            return o1.getId().compareTo(o2.getId());
```

Comment:
We might revisit this when working in improving the LDAP group mapper as we might want to be able to set a specific order when there's more than one mapper (e.g. 2 mappers mapping different group DNs into different group paths in Keycloak). But it is a topic for a separate discussion in the future.

Reply:
Let's consider this a fallback to ensure stable sorting. If we have criteria that are of a higher priority, we can always use those before using this fallback.